### PR TITLE
New Fact Metric UI

### PR DESCRIPTION
**Pull request by jdorn**
### Features and Changes

- Redesign the fact metric modal with a live SQL preview
- Redesign the fact metric page with a new DataList component instead of SQL for better readability
- Hide capping settings for proportion metrics
- Fix bug when switching metric type to quantile while `COUNT(*)` is selected
- Fix broken dev HMR (hot module reloading) on metric modal and page

### Screenshots

New modal:

image

New metric page:

image

### SQL Preview

There are tons of variables that affect what SQL preview we generate.  This documents some of the different combos.

Simple proportion metric (showing both metric and experiment query):
```sql
SELECT
  -- All of the Fact Table's identifier types are supported
  `user_id` AS user,
  -- Each matching user counts as 1 conversion
  1 AS value
FROM
  `Orders`
WHERE
  -- Only after seeing the experiment
  timestamp > exposure_date
GROUP BY user;

SELECT
  variation,
  -- Number of users who converted
  SUM(m.value) as numerator,
  -- Number of users in experiment
  COUNT(*) as denominator,
  -- Final result
  numerator / denominator as value
FROM
  experiment_users u
  LEFT JOIN metric m ON (m.user = u.user)
GROUP BY variation
```

Adding a conversion window (only changes 1st query):
```sql
...
WHERE
  -- Only after seeing the experiment
  timestamp > exposure_date AND
  -- Conversion Metric Window
  timestamp < exposure_date + '3 days'
GROUP BY user
```

Adding a lookback window (only changes 1st query)
```sql
...
WHERE
  -- Only after seeing the experiment
  timestamp > exposure_date AND
  -- Lookback Metric Window
  timestamp > NOW() - '3 days'
GROUP BY user
```

Adding a filter (only changes 1st query):
```sql
...
WHERE
  -- Filter: High Value Order
  price_category = 'high' AND
  -- Only after seeing the experiment
  timestamp > exposure_date
```

Adding an inline filter (only changes 1st query):
```sql
...
WHERE
  event_type = 'Save Feature Rule' AND
  -- Only after seeing the experiment
  timestamp > exposure_date
GROUP BY user
```

Adding multiple inline filters (only changes 1st query):
```sql
...
WHERE
  event_type IN (
    'Save Feature Rule',
    'App Load'
  ) AND
  -- Only after seeing the experiment
  timestamp > exposure_date
GROUP BY user
```

Mean metric with "Count of Rows" value (only changes select fields in both queries):

```sql
SELECT
  `user_id` AS user,
  COUNT(*) AS value
...

SELECT
  variation,
  -- Total metric value
  SUM(m.value) as numerator,
```

Mean metric with numeric column as value (only changes select of 1st query):

```sql
SELECT
  -- All of the Fact Table's identifier types are supported
  `user_id` AS user,
  SUM(amount) AS value
```

Ratio (only 2nd query changes):
```sql
SELECT
  variation,
  -- Total numerator value
  SUM(m.value) as numerator,
  -- Total denominator value
  SUM(d.value) as denominator,
  -- Final result
  numerator / denominator as value
FROM
  experiment_users u
  LEFT JOIN numerator m ON (m.user = u.user)
  LEFT JOIN denominator d ON (d.user = u.user)
GROUP BY variation
```

Event Quantile (both queries have lots of changes):

```sql
SELECT
  -- All of the Fact Table's identifier types are supported
  `user_id` AS user,
  `amount` AS value
FROM
  `Orders`
WHERE
  -- Only after seeing the experiment
  timestamp > exposure_date

SELECT
  variation,
  -- Final result
  PERCENTILE(
    -- COALESCE to include NULL in the calculation
    COALESCE(m.value,0)
  , 0.5) as value
FROM
  experiment_users u
  LEFT JOIN metric m ON (m.user = u.user)
GROUP BY variation
```

Event Quantile with "ignore zeros" (only changes 1st query)
```sql
...
WHERE
  -- Only after seeing the experiment
  timestamp > exposure_date AND
  -- Ignore zeros in percentile
  value > 0
```

User Quantile (both queries change sunstantially):
```sql
SELECT
  -- All of the Fact Table's identifier types are supported
  `user_id` AS user,
  COUNT(*) AS value
FROM
  `Orders`
WHERE
  -- Only after seeing the experiment
  timestamp > exposure_date
GROUP BY user

SELECT
  variation,
  -- Final result
  PERCENTILE(
    -- COALESCE to include NULL in the calculation
    COALESCE(m.value,0)
  , 0.5) as value
FROM
  experiment_users u
  LEFT JOIN metric m ON (m.user = u.user)
GROUP BY variation
```

User Quantile with Ignore Zeros (only 1st query changes):
```sql
...
GROUP BY user
-- Ignore zeros in percentile
HAVING SUM(amount) > 0
```

### CODE REVIEW

File: packages/front-end/components/FactTables/FactMetricModal.tsx [hunk 1]
```diff
@@ -405,92 +423,254 @@ function ColumnRefSelector({
...
+        )
+      : [];
+
+  whereParts.push(
+    `-- Only after seeing the experiment\ntimestamp > exposure_date`
+  );
+
+  if (windowSettings.type === "lookback") {
+    whereParts.push(
+      `-- Lookback Metric Window\ntimestamp > NOW() - '${windowSettings.windowValue} ${windowSettings.windowUnit}'`
```
Comment (lukesonnet) [inline packages/front-end/components/FactTables/FactMetricModal.tsx:468]:
```suggestion
      `-- Lookback Metric Window\ntimestamp > (NOW() - '${windowSettings.windowValue} ${windowSettings.windowUnit}')`
```

Reflects order of operations in SQL and is clearer IMO

File: packages/front-end/components/FactTables/FactMetricModal.tsx [hunk 2]
```diff
@@ -405,92 +424,254 @@ function ColumnRefSelector({
...
+      ? getColumnRefWhereClause(
+          factTable,
+          columnRef,
+          (s) => s.replace(/'/g, "''"),
+          true
+        )
+      : [];
+
+  whereParts.push(
+    `-- Only after seeing the experiment\ntimestamp > exposure_date`
```
Comment (lukesonnet) [inline packages/front-end/components/FactTables/FactMetricModal.tsx:464]:
```suggestion
    `-- Only after seeing the experiment\ntimestamp > exposure_timestamp`
```

File: packages/front-end/components/FactTables/FactMetricModal.tsx [hunk 3]
```diff
@@ -405,92 +424,254 @@ function ColumnRefSelector({
...
+  );
+
+  if (windowSettings.type === "lookback") {
+    whereParts.push(
+      `-- Lookback Metric Window\ntimestamp > (NOW() - '${windowSettings.windowValue} ${windowSettings.windowUnit}')`
+    );
+  } else if (windowSettings.type === "conversion") {
+    whereParts.push(
+      `-- Conversion Metric Window\ntimestamp < (exposure_date + '${windowSettings.windowValue} ${windowSettings.windowUnit}')`
```
Comment (lukesonnet) [inline packages/front-end/components/FactTables/FactMetricModal.tsx:473]:
```suggestion
      `-- Conversion Metric Window\ntimestamp < (exposure_timestamp + '${windowSettings.windowValue} ${windowSettings.windowUnit}')`
```

**Review by lukesonnet — APPROVED** [review]

File: packages/front-end/components/FactTables/FactMetricModal.tsx [hunk 4]
```diff
@@ -405,92 +424,254 @@ function ColumnRefSelector({
...
+          denominator?.column === "$$distinctusers"
+            ? `Number of users who converted`
+            : `Total denominator value`
+        }\n  SUM(d.value)`
+      : `-- Number of users in experiment\n  COUNT(*)`
+  } as denominator,\n  `
+      : ""
+  }${
+    type === "quantile"
+      ? `-- Final result\n  PERCENTILE(${
```
Comment (lukesonnet) [inline packages/front-end/components/FactTables/FactMetricModal.tsx:578]:
```suggestion
      ? `-- Final result\n  APPROX_PERCENTILE(${
```

Just to ward off some questions here, maybe, and be more accurate. Feel free to ignore.

Comment (jdorn) [reply packages/front-end/components/FactTables/FactMetricModal.tsx:578]: Doesn't seem to be any standardization here between engines.  I think I'll leave it as just `PERCENTILE`.
- BigQuery: `APPROX_QUANTILES`
- ClickHouse: `quantile`
- MS SQL: `APPROX_PERCENTILE_CONT`
- Postgres: `PERCENTILE_CONT`
- Vertica: `APPROXIMATE_PERCENTILE`

Comment (lukesonnet) [reply packages/front-end/components/FactTables/FactMetricModal.tsx:578]: They are all approximations except for postgres, AFAIK.

**Pull request merged**